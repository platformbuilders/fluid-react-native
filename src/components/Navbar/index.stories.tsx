import type { Meta } from '@storybook/react';
import Navbar from '.';

const meta: Meta<typeof Navbar> = {
  title: 'Navbar',
  component: Navbar,
  argTypes: {},
  args: {
    activeFieldColor: '#fff',
    currentRoute: 'Home',
    fields: [
      {
        label: 'Home',
        iconName: 'home',
        onPress: () => console.log('Home'),
        routeName: 'Home',
      },
      {
        label: 'Login',
        iconName: 'user',
        onPress: () => console.log('Login'),
        routeName: 'Login',
      },
    ],
  },
  parameters: {
    notes: 'Este componente é utilizado para renderizar a barra de navegação.',
  },
};

export default meta;

export const Default = {};
