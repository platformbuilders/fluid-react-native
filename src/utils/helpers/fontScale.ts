import { RFValue } from 'react-native-responsive-fontsize';

export const scaledFontSize = (
  size: number | string,
  baselineFactor?: 280,
): number => {
  const value = typeof size === 'string' ? parseInt(size, 10) : size;
  return RFValue(value, baselineFactor);
};
