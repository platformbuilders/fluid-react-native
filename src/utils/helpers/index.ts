export {
  currencyParser,
  parseToThousands,
  removeWhiteSpaces,
  toOnlyNumbers,
  getTheme,
  ifStyle,
  switchStyle,
} from '@platformbuilders/helpers';
export * from './generateHaptic';
export * from './getByTypographyVariant';
export * from './fontScale';
export * from './shadow';
