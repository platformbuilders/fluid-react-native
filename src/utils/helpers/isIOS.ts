import { Platform } from 'react-native';

const IOS_PLATFORM = 'ios';

export const isIOS = (): boolean => Platform.OS === IOS_PLATFORM;
