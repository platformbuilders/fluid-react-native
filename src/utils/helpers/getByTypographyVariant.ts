import { getTheme } from '@platformbuilders/helpers';
import { scaledFontSize } from './fontScale';

export const getFontSize = (props: any): number => {
  const fontSize = getTheme(`${props.variant}.fontSize`)(props) || 16;
  return scaledFontSize(fontSize);
};

export const getLineHeight = (props: any): number => {
  const lineHeight = getTheme(`${props.variant}.lineHeight`)(props) || 22;
  return scaledFontSize(lineHeight);
};
