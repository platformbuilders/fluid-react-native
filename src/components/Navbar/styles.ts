import { moderateScale } from 'react-native-size-matters';
import styled from 'styled-components/native';
import { ThemeProps, getTheme, ifStyle } from '@platformbuilders/theme-toolkit';
import { getShadow } from '../../utils/helpers';
import Icon from '../Icon';
import Touchable from '../Touchable';
import Typography from '../Typography';

const textColor = getTheme('text.main');
const primaryMain = getTheme('brand.primary.main');
const backgroundZ4 = getTheme('background.z4');

const emptyOpacity = getTheme('opacity.empty');
const opaqueOpacity = getTheme('opacity.opaque');

const smSpacing = getTheme('spacing.sm');

const isActive = ifStyle('active');

type NavbarWrapperProps = {
  safeBottomArea: boolean;
  bottomSpacing: number;
};

type FieldsProps = {
  active: boolean;
  activeColor?: string;
} & ThemeProps;

export const NavbarWrapper = styled.View<NavbarWrapperProps>`
  width: 100%;
  padding: ${smSpacing}px;
  padding-bottom: ${({ bottomSpacing }) => `${bottomSpacing}px`};
  background-color: ${backgroundZ4};
  flex-direction: row;
  justify-content: space-evenly;
  align-items: flex-start;
  margin-top: auto;
  ${getShadow()};
`;

export const NavField = styled(Touchable)`
  width: ${moderateScale(70)}px;
  justify-content: center;
  align-items: center;
`;

export const NavIcon = styled(Icon).attrs((props: FieldsProps) => ({
  size: moderateScale(30),
  color: isActive(
    props?.activeColor || primaryMain(props),
    textColor(props),
  )(props),
  touchable: false,
}))<FieldsProps>`
  opacity: ${isActive(opaqueOpacity, emptyOpacity)};
`;

export const NavLabel = styled(Typography).attrs({
  variant: 'xs',
})<FieldsProps>`
  color: ${({ activeColor }: FieldsProps) =>
    isActive(activeColor || primaryMain, textColor)};
  text-transform: uppercase;
  font-weight: bold;
  text-align: center;
  opacity: ${isActive(opaqueOpacity, emptyOpacity)};
`;
