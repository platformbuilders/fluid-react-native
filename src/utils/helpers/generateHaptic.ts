import Haptic, { HapticFeedbackType } from 'react-native-haptic';
import { isIOS } from './isIOS';

export const generateHaptic = (haptic: HapticFeedbackType = 'impact'): void => {
  if (isIOS()) {
    Haptic.generate(haptic);
  }
};
