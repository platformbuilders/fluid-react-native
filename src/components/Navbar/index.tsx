import React, { useCallback } from 'react';
import { NavField, NavIcon, NavLabel, NavbarWrapper } from './styles';

type Props = {
  fields: {
    routeName: string;
    onPress(): void;
    iconName: string;
    label: string;
  }[];
  currentRoute: string;
  activeFieldColor?: string;
  activateSafeBottomArea?: boolean;
};

const Navbar: React.FC<Props> = ({
  fields,
  currentRoute,
  activeFieldColor,
  activateSafeBottomArea,
}) => {
  const renderFields = useCallback(
    () =>
      fields.map((navField) => (
        <NavField
          key={navField.routeName}
          accessibility={`Ir para ${navField.label}`}
          onPress={navField.onPress}
          id={navField.routeName}
          {...(navField as any)}
        >
          <NavIcon
            accessibility={navField.iconName}
            name={navField.iconName}
            active={navField.routeName === currentRoute}
            activeColor={activeFieldColor}
            id={`Icon${navField.routeName}`}
          />

          <NavLabel
            active={navField.routeName === currentRoute}
            activeColor={activeFieldColor}
            id={`Label${navField.routeName}`}
          >
            {navField.label}
          </NavLabel>
        </NavField>
      )),
    [fields, currentRoute],
  );

  return (
    <NavbarWrapper safeBottomArea={!!activateSafeBottomArea}>
      {renderFields()}
    </NavbarWrapper>
  );
};

export default Navbar;
