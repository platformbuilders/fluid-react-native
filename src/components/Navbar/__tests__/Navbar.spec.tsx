import React from 'react';
import { fireEvent, render } from 'react-native-testing-library';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';
import Navbar from '..';
import { theme } from '../../../test/helpers';

const onPressEvent = jest.fn();

const fields = [
  {
    routeName: 'Home',
    onPress: onPressEvent,
    iconName: 'flag',
    label: 'Home',
  },
];

describe('<Navbar />', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render navbar component', () => {
    const wrapper = renderer.create(
      <ThemeProvider theme={theme}>
        <Navbar currentRoute="Home" fields={fields} />
      </ThemeProvider>,
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should trigger onPress function when nav field is pressed', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Navbar currentRoute="Home" fields={fields} />
      </ThemeProvider>,
    );

    const component = getByTestId('Home');

    fireEvent.press(component);

    expect(onPressEvent).toHaveBeenCalled();
  });

  it('should render navbar component with custom color', () => {
    const wrapper = renderer.create(
      <ThemeProvider theme={theme}>
        <Navbar
          currentRoute="Home"
          fields={fields}
          activeFieldColor={'#000000'}
        />
      </ThemeProvider>,
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render navbar component with safe bottom area', () => {
    const wrapper = renderer.create(
      <ThemeProvider theme={theme}>
        <Navbar currentRoute="Home" fields={fields} activateSafeBottomArea />
      </ThemeProvider>,
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
