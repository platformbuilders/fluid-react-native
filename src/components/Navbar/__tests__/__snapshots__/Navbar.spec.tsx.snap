// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> should render navbar component 1`] = `
<View
  safeBottomArea={false}
  style={
    Array [
      Object {
        "alignItems": "flex-start",
        "backgroundColor": "#fff",
        "elevation": 3,
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "marginTop": "auto",
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 6.285714285714286,
        "width": "100%",
      },
    ]
  }
>
  <View
    accessibilityLabel="Ir para Home"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "width": 110,
      }
    }
    testID="Home"
  >
    <View
      collapsable={false}
      nativeID="animatedComponent"
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        accessibilityLabel="flag"
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="IconHome"
      >
        <View>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 47.14285714285714,
                },
                undefined,
                Object {
                  "fontFamily": "FontAwesome5Free-Regular",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "400",
                },
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
    <Text
      accessibilityLabel="Home"
      active={true}
      lineHeightVariant="min"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 13,
            "lineHeight": 13,
          },
          Object {
            "color": "#007AFF",
            "fontWeight": "bold",
            "opacity": 1,
            "textAlign": "center",
            "textTransform": "uppercase",
          },
        ]
      }
      testID="LabelHome"
      variant="xs"
    >
      Home
    </Text>
  </View>
</View>
`;

exports[`<Navbar /> should render navbar component with custom color 1`] = `
<View
  safeBottomArea={false}
  style={
    Array [
      Object {
        "alignItems": "flex-start",
        "backgroundColor": "#fff",
        "elevation": 3,
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "marginTop": "auto",
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 6.285714285714286,
        "width": "100%",
      },
    ]
  }
>
  <View
    accessibilityLabel="Ir para Home"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "width": 110,
      }
    }
    testID="Home"
  >
    <View
      collapsable={false}
      nativeID="animatedComponent"
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        accessibilityLabel="flag"
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="IconHome"
      >
        <View>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontSize": 47.14285714285714,
                },
                undefined,
                Object {
                  "fontFamily": "FontAwesome5Free-Regular",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "400",
                },
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
    <Text
      accessibilityLabel="Home"
      active={true}
      activeColor="#000000"
      lineHeightVariant="min"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 13,
            "lineHeight": 13,
          },
          Object {
            "color": "#000000",
            "fontWeight": "bold",
            "opacity": 1,
            "textAlign": "center",
            "textTransform": "uppercase",
          },
        ]
      }
      testID="LabelHome"
      variant="xs"
    >
      Home
    </Text>
  </View>
</View>
`;

exports[`<Navbar /> should render navbar component with safe bottom area 1`] = `
<View
  safeBottomArea={true}
  style={
    Array [
      Object {
        "alignItems": "flex-start",
        "backgroundColor": "#fff",
        "elevation": 3,
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "marginTop": "auto",
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 6.285714285714286,
        "width": "100%",
      },
    ]
  }
>
  <View
    accessibilityLabel="Ir para Home"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "width": 110,
      }
    }
    testID="Home"
  >
    <View
      collapsable={false}
      nativeID="animatedComponent"
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        accessibilityLabel="flag"
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="IconHome"
      >
        <View>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 47.14285714285714,
                },
                undefined,
                Object {
                  "fontFamily": "FontAwesome5Free-Regular",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "400",
                },
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
    <Text
      accessibilityLabel="Home"
      active={true}
      lineHeightVariant="min"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 13,
            "lineHeight": 13,
          },
          Object {
            "color": "#007AFF",
            "fontWeight": "bold",
            "opacity": 1,
            "textAlign": "center",
            "textTransform": "uppercase",
          },
        ]
      }
      testID="LabelHome"
      variant="xs"
    >
      Home
    </Text>
  </View>
</View>
`;
